Fix fatal config set error

sepehr-rs opened the pull request:
Fixes #13279 

I tried fixing this by providing the user with a helpful error message instead of crashing, and made a new parser to allow writing to the PIP_CONFIG_FILE environment variable temporarily.

One important caveat is that changes made to the config file via `PIP_CONFIG_FILE` may not be reflected in other commands like `pip config get` due to how pip loads config scopes.

Would appreciate any feedback on this :).

sepehr-rs committed:
Fix user configuration error when trying to change global settings

sepehr-rs committed:
Add news entry file

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Update 13279.bugfix.rst

sepehr-rs committed:
Update configuration.py to avoid pre-commit errors

notatallshaw commented:
> I think pre-commit is showing weird behavior here again...
> I've changed the problematic files but it is still failing with the old files.

Do you know if there is anything particular about your git workflow? It does seem to be happening with your PRs more than anyone else's (though I have seen it happen at least once to mine).

You could perhaps report this to pre-commit, I would normally assist in reporting a bug but my experience with that project means I am inclined to help it out.

In the mean time I'm thinking we should move off pre-commit.ci, and instead just add it as a github workflow, I don't see any advantage to having it run in a separate infrastructure. I'll make an issue and a PR when I have a moment.

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Merge branch 'main' into fix-config-error

notatallshaw commented on src/pip/_internal/configuration.py:
```diff
@@ -378,7 +378,22 @@ def _get_parser_to_modify(self) -> tuple[str, RawConfigParser]:
+                    "as this config file is outside pip's normal config scopes. "
+                    "Please inspect the file manually or unset "
+                    "PIP_CONFIG_FILE for typical behavior.".format(
+                        env_config_file, self.load_only
```
I'm not quite sure I follow what this is trying to say, you can you give me a high level description of what you want to convey to the user.

sepehr-rs replied:
Thank you for reviewing this!
I wanted to let the user know that when the `PIP_CONFIG_FILE` environment variable is set, pip uses that file instead of its normal configuration hierarchy. As a result, some commands may not reflect changes made to this file, since it lies outside pip’s typical config scopes.

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Added a patch for #12706

sepehr-rs committed:
Merge branch 'fix-config-error' of https://github.com/sepehr-rs/pip into fix-config-error

sepehr-rs committed:
Fix precommit

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Merge branch 'main' into fix-config-error

notatallshaw commented:
I don't think this PR fully addresses all the review comments yet, I will try and take a look in a few weeks.

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Change logger warning message

sepehr-rs commented:
Gentle ping. I believe this is ready for review. This PR also addresses #12706, #11880, and #12481. I'd appreciate any feedback when someone has time.

notatallshaw commented:
Thanks for the ping, I'm going to be away on vacation for a week, but I hope some time in June to sit down and do make a big dent in the backlog of things I've promised to review.

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs committed:
Merge branch 'main' into fix-config-error

notatallshaw commented on src/pip/_internal/configuration.py:
```diff
@@ -373,16 +371,38 @@ def get_values_in_config(self, variant: Kind) -> dict[str, Any]:
+
+        section, name = _disassemble_key(key)
+        for fname, parser in reversed(parsers):
+            if parser.has_section(section) and parser.has_option(section, name):
```
The `has_option` search (and `remove_option` earlier in the file) uses the normalized name, so `unset global.no-cache-dir` still hits id=1 when the file has `no_cache_dir` with underscores.

Now a regression, but it's the #12706 case, so the fix is incomplete unless we normalize at the parser layer.

notatallshaw commented on src/pip/_internal/configuration.py:
```diff
@@ -373,16 +371,38 @@ def get_values_in_config(self, variant: Kind) -> dict[str, Any]:
         parsers = self._parsers[self.load_only]
         if not parsers:
-            # This should not happen if everything works correctly.
+            env_config_file = os.environ.get("PIP_CONFIG_FILE")
```
no devnull guard on the write side, so `PIP_CONFIG_FILE=/dev/null pip config set ...` now exits 0 and drops the value. 

#11880 asked for an error, not a warned no-op. (`os.devnull` is `nul` on Windows too.)

notatallshaw commented on src/pip/_internal/configuration.py:
```diff
@@ -373,16 +371,38 @@ def get_values_in_config(self, variant: Kind) -> dict[str, Any]:
+            env_config_file = os.environ.get("PIP_CONFIG_FILE")
+            if env_config_file and self.load_only in (kinds.USER, kinds.GLOBAL):
+                parser = self._construct_parser(env_config_file)
+                self._parsers[kinds.ENV].append((env_config_file, parser))
```
I think this should be append to `_parsers[self.load_only]`, not `_parsers[kinds.ENV]`, or it's not re-entrant.

notatallshaw commented on src/pip/_internal/configuration.py:
```diff
@@ -184,12 +184,10 @@ def unset_value(self, key: str) -> None:
-            key not in self._config[self.load_only][fname]
-            and key not in self._config[self.load_only]
-        ):
+        file_config = self._config[self.load_only].get(fname, {})
```
`unset` never works under `PIP_CONFIG_FILE`, even though `set` now does. `set_value` writes to the env file; `unset_value` hits the same fallback (prints the "changes will be applied" warning) and then this guard fires, because `_config[load_only]` is empty when the file is loaded under ENV. So:

```shell
    $ pip config set   global.x v   # writes, exit 0
    $ pip config unset global.x     # same warning, then "No such key", exit 1
```

Can we point unset at the ENV scope / the parser we just built so it removes the key instead of denying it exists?

notatallshaw reviewed (commented):
The `except IndexError` in `get_file_to_edit`  is now dead, the fallback returns or raises `ConfigurationError`.

Would be good to have some tests for unset + `PIP_CONFIG_FILE`, or some unit or functional tests in general.

sepehr-rs committed:
Address reviews

sepehr-rs committed:
Add tests

sepehr-rs commented:
Thanks so much for the thorough reviews, @notatallshaw!

I've addressed the remaining points, got `unset` working, made the code raise an exception when `PIP_CONFIG_FILE` is set to `devnull`, and added tests for both.

I'd appreciate another look when you have a chance!

sepehr-rs committed:
Merge branch 'main' into fix-config-error

sepehr-rs commented:
Hey @ichard26, would you be open to considering adding this to the upcoming release milestone? I believe the latest review comments have been addressed and the requested tests have been added. No worries if it's too late in the release cycle or you'd prefer to leave it for a later release.

ichard26 commented:
TBH, pip's configuration code is one of my least favourite parts of the codebase to make changes. It's a mess to follow[^1] while being an easy source of backwards compatibility potholes. I'd need to invest a significant amount of time to understand what's the underlying issue here and whether this change is the right solution.

So, I consider this PR out of scope for the upcoming 26.2 release. I'll tack to the 26.3 milestone for the time being.

[^1]: Not critiquing the code itself or those who wrote any of it, but it's impenetrably hard to read
